File: src/vpc/db/migrate.js
"use strict";

const fs = require("fs");
const path = require("path");
const { Pool } = require("pg");
const { requireAdmin, requireSuperAdmin } = require("../../common/auth");
const { getHeader } = require("../../common/validate");

let _pool;

function pool() {
  if (_pool) return _pool;

  const { PGHOST, PGPORT, PGDATABASE, PGUSER, PGPASSWORD, PGSSLMODE } = process.env;
  if (!PGHOST || !PGDATABASE || !PGUSER || !PGPASSWORD) return null;

  _pool = new Pool({
    host: PGHOST,
    port: Number(PGPORT || 5432),
    database: PGDATABASE,
    user: PGUSER,
    password: PGPASSWORD,
    ssl: (PGSSLMODE || "").toLowerCase() === "disable" ? false : { rejectUnauthorized: false },
    max: Number(process.env.PG_MAX || 1), // ✅ Lambda 推荐 1
    connectionTimeoutMillis: Number(process.env.PG_CONNECTION_TIMEOUT || 5000),
    idleTimeoutMillis: Number(process.env.PG_IDLE_TIMEOUT || 30000),
  });

  return _pool;
}

function loadMigrations() {
  const dir = path.join(__dirname, "migrations");
  if (!fs.existsSync(dir)) return [];

  const files = fs
    .readdirSync(dir)
    .filter((f) => /^\d+_.+\.sql$/i.test(f))
    .sort((a, b) => a.localeCompare(b, "en"));

  const migrations = files.map((file) => {
    const base = file.replace(/\.sql$/i, "");
    const versionStr = base.split("_")[0];
    const version = Number.parseInt(versionStr, 10);
    if (!Number.isFinite(version)) throw new Error(`Bad migration filename (no numeric prefix): ${file}`);
    return { version, name: base, file, fullPath: path.join(dir, file) };
  });

  // check duplicates
  const seen = new Set();
  for (const m of migrations) {
    if (seen.has(m.version)) throw new Error(`Duplicate migration version: ${m.version}`);
    seen.add(m.version);
  }

  return migrations;
}

async function ensureMigrationsTable(client) {
  await client.query(`
    create table if not exists schema_migrations (
      version int primary key,
      name text not null,
      applied_at timestamptz not null default now()
    );
  `);
}

async function getAppliedVersions(client) {
  await ensureMigrationsTable(client);
  const r = await client.query(`select version from schema_migrations order by version asc;`);
  return new Set(r.rows.map((x) => Number(x.version)));
}

async function applyOne(client, m) {
  const sql = fs.readFileSync(m.fullPath, "utf8");

  await client.query("BEGIN");
  try {
    // 执行 SQL（DDL 允许多语句）
    await client.query(sql);

    // 记录 migration
    await client.query(
      `insert into schema_migrations(version, name) values ($1, $2) on conflict (version) do nothing;`,
      [m.version, m.name]
    );

    await client.query("COMMIT");
  } catch (e) {
    await client.query("ROLLBACK");
    throw e;
  }
}

async function withMigrationLock(client, fn) {
  // 固定 lock id（随便一个稳定整数即可）
  const LOCK_ID = 77889911;
  await client.query(`select pg_advisory_lock($1);`, [LOCK_ID]);
  try {
    return await fn();
  } finally {
    try {
      await client.query(`select pg_advisory_unlock($1);`, [LOCK_ID]);
    } catch {
      // ignore
    }
  }
}

async function handleDbPing({ res, auth }) {
  const okRes = requireAdmin({ auth, res });
  if (okRes !== true) return okRes;

  const p = pool();
  if (!p) return res.badRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

  const r = await p.query("select 1 as ok;");
  return res.ok({ ok: r.rows[0].ok === 1 });
}

async function handleDbMigrate({ event, res, auth, query }) {
  const okRes = requireSuperAdmin({ auth, res });
  if (okRes !== true) return okRes;

  // 额外手动保护：必须带 x-migrate-secret（如果 env 配了）
  const required = process.env.MIGRATE_SECRET || "";
  if (required) {
    const got = String(getHeader(event, "x-migrate-secret") || "");
    if (got !== required) return res.forbidden("Bad migrate secret");
  }

  const p = pool();
  if (!p) return res.badRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

  const migrations = loadMigrations();
  if (migrations.length === 0) return res.badRequest("MIGRATIONS_EMPTY", "No migrations found in src/vpc/db/migrations");

  const dryRun = String(query.dryRun || "").toLowerCase() === "true";

  const client = await p.connect();
  try {
    const result = await withMigrationLock(client, async () => {
      const applied = await getAppliedVersions(client);
      const pending = migrations.filter((m) => !applied.has(m.version));

      if (dryRun) {
        return {
          dryRun: true,
          available: migrations.map((m) => ({ version: m.version, name: m.name, file: m.file })),
          pending: pending.map((m) => ({ version: m.version, name: m.name, file: m.file })),
        };
      }

      const appliedNow = [];
      for (const m of pending) {
        await applyOne(client, m);
        appliedNow.push({ version: m.version, name: m.name, file: m.file });
      }

      return {
        dryRun: false,
        applied: appliedNow,
        appliedCount: appliedNow.length,
        latestAvailable: migrations[migrations.length - 1].version,
      };
    });

    return res.ok(result);
  } finally {
    client.release();
  }
}

async function handleDbMigrationsList({ res, auth }) {
  const okRes = requireSuperAdmin({ auth, res });
  if (okRes !== true) return okRes;

  const p = pool();
  if (!p) return res.badRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

  const migrations = loadMigrations();
  const client = await p.connect();
  try {
    await ensureMigrationsTable(client);
    const r = await client.query(`select version, name, applied_at from schema_migrations order by version asc;`);

    const applied = r.rows.map((x) => ({
      version: Number(x.version),
      name: x.name,
      appliedAt: x.applied_at ? new Date(x.applied_at).toISOString() : null,
    }));

    const appliedSet = new Set(applied.map((x) => x.version));
    const pending = migrations
      .filter((m) => !appliedSet.has(m.version))
      .map((m) => ({ version: m.version, name: m.name, file: m.file }));

    return res.ok({
      available: migrations.map((m) => ({ version: m.version, name: m.name, file: m.file })),
      applied,
      pending,
    });
  } finally {
    client.release();
  }
}

async function handleDbListDatabases({ res, auth }) {
  const okRes = requireSuperAdmin({ auth, res });
  if (okRes !== true) return okRes;

  const p = pool();
  if (!p) return res.badRequest("CONFIG_ERROR", "Missing PG env vars");

  const r = await p.query(`
    SELECT datname
    FROM pg_database
    WHERE datistemplate = false
    ORDER BY datname;
  `);

  return res.ok({ databases: r.rows.map(x => x.datname) });
}
function isValidDbName(name) {
  return /^[a-zA-Z_][a-zA-Z0-9_]{0,62}$/.test(name);
}

async function handleDbCreateDatabase({ event, res, auth, query }) {
  const okRes = requireSuperAdmin({ auth, res });
  if (okRes !== true) return okRes;

  // 额外保护：必须带 migrate secret（建议）
  const required = process.env.MIGRATE_SECRET || "";
  if (required) {
    const got = String(getHeader(event, "x-migrate-secret") || "");
    if (got !== required) return res.forbidden("Bad migrate secret");
  }

  // 强制要求当前连接在 postgres 维护库上
  if (String(process.env.PGDATABASE || "") !== "postgres") {
    return res.badRequest("CONFIG_ERROR", "PGDATABASE must be 'postgres' for CREATE DATABASE");
  }

  const name = String(query.name || "").trim();
  if (!name) return res.badRequest("BAD_REQUEST", "Missing query param: ?name=");
  if (!isValidDbName(name)) return res.badRequest("BAD_REQUEST", "Bad database name");

  // CREATE DATABASE 不能放事务里——这里直接执行即可
  const exists = await pool().query("SELECT 1 FROM pg_database WHERE datname=$1", [name]);
  if (exists.rowCount === 0) {
    await pool().query(`CREATE DATABASE ${name};`);
    return res.ok({ ok: true, created: name });
  }

  return res.ok({ ok: true, existed: name });
}
async function handleDbDropAndRecreate({ event, res, auth, query }) {
  const okRes = requireSuperAdmin({ auth, res });
  if (okRes !== true) return okRes;

  // extra manual guard (same as migrate)
  const required = process.env.MIGRATE_SECRET || "";
  if (required) {
    const got = String(getHeader(event, "x-migrate-secret") || "");
    if (got !== required) return res.forbidden("Bad migrate secret");
  }

  // MUST run on maintenance DB
  if (String(process.env.PGDATABASE || "") !== "postgres") {
    return res.badRequest("CONFIG_ERROR", "PGDATABASE must be 'postgres' to drop/create databases");
  }

  const name = String(query.name || "").trim();
  if (!name) return res.badRequest("BAD_REQUEST", "Missing query param: ?name=");
  if (!isValidDbName(name)) return res.badRequest("BAD_REQUEST", "Bad database name");
  if (name === "postgres" || name === "rdsadmin") {
    return res.badRequest("BAD_REQUEST", "Refusing to drop system database");
  }

  const p = pool();
  if (!p) return res.badRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

  // 1) terminate existing connections
  //    (RDS master can do this; ignore if none)
  await p.query(
    `
    SELECT pg_terminate_backend(pid)
    FROM pg_stat_activity
    WHERE datname = $1
      AND pid <> pg_backend_pid();
  `,
    [name]
  );

    try {
      await p.query(`DROP DATABASE IF EXISTS ${name} WITH (FORCE);`);
      await p.query(`CREATE DATABASE ${name};`);
      return res.ok({ ok: true, recreated: name });
    } catch (e) {
      // 把 PG 原始错误 message 返回，方便你定位
      return res.badRequest("DB_RECREATE_FAILED", String(e?.message || e));
    }
}
module.exports = {
  handleDbPing,
  handleDbMigrate,
  handleDbMigrationsList,
  handleDbCreateDatabase,
  handleDbListDatabases,
  handleDbDropAndRecreate
};

File: src_C/Vpc/Db/Migrate.cs
using System.Globalization;
using Npgsql;
using RecallSmith.Lambda.Common;
using static RecallSmith.Lambda.Vpc.Db.DbUtil;

namespace RecallSmith.Lambda.Vpc.Db;

public static class Migrate
{
  private sealed record Migration(int Version, string Name, string File, string FullPath);

  private static string MigrationsDir()
  {
    // Copied to output via csproj <CopyToOutputDirectory>.
    return Path.Combine(AppContext.BaseDirectory, "Db", "Migrations");
  }

  private static List<Migration> LoadMigrations()
  {
    var dir = MigrationsDir();
    if (!Directory.Exists(dir)) return [];

    var files = Directory
      .EnumerateFiles(dir, "*.sql", SearchOption.TopDirectoryOnly)
      .Select(f => Path.GetFileName(f)!)
      .Where(f => System.Text.RegularExpressions.Regex.IsMatch(f, "^\\d+_.+\\.sql$", System.Text.RegularExpressions.RegexOptions.IgnoreCase))
      .OrderBy(f => f, StringComparer.Ordinal)
      .ToList();

    var migrations = new List<Migration>();
    foreach (var file in files)
    {
      var baseName = file[..^4]; // .sql
      var versionStr = baseName.Split('_')[0];
      if (!int.TryParse(versionStr, NumberStyles.Integer, CultureInfo.InvariantCulture, out var version))
      {
        throw new InvalidOperationException($"Bad migration filename (no numeric prefix): {file}");
      }

      migrations.Add(new Migration(version, baseName, file, Path.Combine(dir, file)));
    }

    var seen = new HashSet<int>();
    foreach (var m in migrations)
    {
      if (!seen.Add(m.Version)) throw new InvalidOperationException($"Duplicate migration version: {m.Version}");
    }

    return migrations;
  }

  private static async Task EnsureMigrationsTable(NpgsqlConnection conn, NpgsqlTransaction? tx)
  {
    const string sql = """
      create table if not exists schema_migrations (
        version int primary key,
        name text not null,
        applied_at timestamptz not null default now()
      );
      """;

    await ExecuteAsync(conn, tx, sql, []);
  }

  private static async Task<HashSet<int>> GetAppliedVersions(NpgsqlConnection conn, NpgsqlTransaction? tx)
  {
    await EnsureMigrationsTable(conn, tx);
    const string sql = "select version from schema_migrations order by version asc;";
    var rows = await QueryAsync(conn, tx, sql, []);
    return rows.Select(r => Convert.ToInt32(r["version"], CultureInfo.InvariantCulture)).ToHashSet();
  }

  private static async Task ApplyOne(NpgsqlConnection conn, Migration m)
  {
    var sql = await File.ReadAllTextAsync(m.FullPath);

    await using var tx = await conn.BeginTransactionAsync();
    try
    {
      await ExecuteAsync(conn, tx, sql, []);
      await ExecuteAsync(
        conn,
        tx,
        "insert into schema_migrations(version, name) values ($1, $2) on conflict (version) do nothing;",
        [m.Version, m.Name]);

      await tx.CommitAsync();
    }
    catch
    {
      try { await tx.RollbackAsync(); } catch { /* ignore */ }
      throw;
    }
  }

  private static async Task<T> WithMigrationLock<T>(NpgsqlConnection conn, Func<Task<T>> fn)
  {
    const int lockId = 77889911;
    await ExecuteAsync(conn, null, "select pg_advisory_lock($1);", [lockId]);
    try
    {
      return await fn();
    }
    finally
    {
      try { await ExecuteAsync(conn, null, "select pg_advisory_unlock($1);", [lockId]); } catch { /* ignore */ }
    }
  }

  public static async Task<Amazon.Lambda.APIGatewayEvents.APIGatewayProxyResponse> HandleDbPing(
    LambdaRequest req,
    Res res,
    AuthContext auth)
  {
    var deny = Auth.RequireAdmin(auth, res);
    if (deny is not null) return deny;

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null)
    {
      return res.BadRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");
    }

    var scalar = await ExecuteScalarAsync(conn, null, "select 1 as ok;", []);
    var ok = scalar is not null && Convert.ToInt32(scalar, CultureInfo.InvariantCulture) == 1;
    return res.Ok(new { ok });
  }

  public static async Task<Amazon.Lambda.APIGatewayEvents.APIGatewayProxyResponse> HandleDbMigrate(
    LambdaRequest req,
    Res res,
    AuthContext auth)
  {
    var deny = Auth.RequireSuperAdmin(auth, res);
    if (deny is not null) return deny;

    // extra manual guard: x-migrate-secret (if configured)
    var required = Environment.GetEnvironmentVariable("MIGRATE_SECRET") ?? string.Empty;
    if (!string.IsNullOrEmpty(required))
    {
      var got = Validation.GetHeader(req, "x-migrate-secret") ?? string.Empty;
      if (!string.Equals(got, required, StringComparison.Ordinal)) return res.Forbidden("Bad migrate secret");
    }

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null)
    {
      return res.BadRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");
    }

    var migrations = LoadMigrations();
    if (migrations.Count == 0)
    {
      return res.BadRequest("MIGRATIONS_EMPTY", "No migrations found in Db/Migrations");
    }

    var dryRun = string.Equals(req.Query.TryGetValue("dryRun", out var d) ? d : null, "true", StringComparison.OrdinalIgnoreCase);

    var result = await WithMigrationLock<object>(conn, async () =>
    {
      var applied = await GetAppliedVersions(conn, null);
      var pending = migrations.Where(m => !applied.Contains(m.Version)).ToList();

      if (dryRun)
      {
        return (object)new
        {
          dryRun = true,
          available = migrations.Select(m => new { version = m.Version, name = m.Name, file = m.File }).ToList(),
          pending = pending.Select(m => new { version = m.Version, name = m.Name, file = m.File }).ToList(),
        };
      }

      var appliedNow = new List<object>();
      foreach (var m in pending)
      {
        await ApplyOne(conn, m);
        appliedNow.Add(new { version = m.Version, name = m.Name, file = m.File });
      }

      return (object)new
      {
        dryRun = false,
        applied = appliedNow,
        appliedCount = appliedNow.Count,
        latestAvailable = migrations[^1].Version,
      };
    });

    return res.Ok(result);
  }

  public static async Task<Amazon.Lambda.APIGatewayEvents.APIGatewayProxyResponse> HandleDbMigrationsList(
    LambdaRequest req,
    Res res,
    AuthContext auth)
  {
    var deny = Auth.RequireSuperAdmin(auth, res);
    if (deny is not null) return deny;

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null) return res.BadRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

    var migrations = LoadMigrations();

    await EnsureMigrationsTable(conn, null);
    var rows = await QueryAsync(conn, null, "select version, name, applied_at from schema_migrations order by version asc;", []);

    var applied = rows.Select(x => new
    {
      version = Convert.ToInt32(x["version"], CultureInfo.InvariantCulture),
      name = Convert.ToString(x["name"], CultureInfo.InvariantCulture),
      appliedAt = x["applied_at"] is DateTime dt ? new DateTimeOffset(dt).ToString("O") : null,
    }).ToList();

    var appliedSet = applied.Select(x => x.version).ToHashSet();
    var pending = migrations
      .Where(m => !appliedSet.Contains(m.Version))
      .Select(m => new { version = m.Version, name = m.Name, file = m.File })
      .ToList();

    return res.Ok(new
    {
      available = migrations.Select(m => new { version = m.Version, name = m.Name, file = m.File }).ToList(),
      applied,
      pending,
    });
  }

  public static async Task<Amazon.Lambda.APIGatewayEvents.APIGatewayProxyResponse> HandleDbListDatabases(
    LambdaRequest req,
    Res res,
    AuthContext auth)
  {
    var deny = Auth.RequireSuperAdmin(auth, res);
    if (deny is not null) return deny;

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null) return res.BadRequest("CONFIG_ERROR", "Missing PG env vars");

    const string sql = """
      SELECT datname
      FROM pg_database
      WHERE datistemplate = false
      ORDER BY datname;
      """;

    var rows = await QueryAsync(conn, null, sql, []);
    return res.Ok(new { databases = rows.Select(x => x["datname"]).ToList() });
  }

  public static async Task<Amazon.Lambda.APIGatewayEvents.APIGatewayProxyResponse> HandleDbCreateDatabase(
    LambdaRequest req,
    Res res,
    AuthContext auth)
  {
    var deny = Auth.RequireSuperAdmin(auth, res);
    if (deny is not null) return deny;

    var required = Environment.GetEnvironmentVariable("MIGRATE_SECRET") ?? string.Empty;
    if (!string.IsNullOrEmpty(required))
    {
      var got = Validation.GetHeader(req, "x-migrate-secret") ?? string.Empty;
      if (!string.Equals(got, required, StringComparison.Ordinal)) return res.Forbidden("Bad migrate secret");
    }

    if (!string.Equals(Environment.GetEnvironmentVariable("PGDATABASE") ?? string.Empty, "postgres", StringComparison.Ordinal))
    {
      return res.BadRequest("CONFIG_ERROR", "PGDATABASE must be 'postgres' for CREATE DATABASE");
    }

    var name = (req.Query.TryGetValue("name", out var n) ? n : string.Empty).Trim();
    if (string.IsNullOrEmpty(name)) return res.BadRequest("BAD_REQUEST", "Missing query param: ?name=");
    if (!Validation.IsValidDbName(name)) return res.BadRequest("BAD_REQUEST", "Bad database name");

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null) return res.BadRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

    var exists = await QueryAsync(conn, null, "SELECT 1 FROM pg_database WHERE datname=$1", [name]);
    if (exists.Count == 0)
    {
      await ExecuteAsync(conn, null, $"CREATE DATABASE {name};", []);
      return res.Ok(new { ok = true, created = name });
    }

    return res.Ok(new { ok = true, existed = name });
  }

  public static async Task<Amazon.Lambda.APIGatewayEvents.APIGatewayProxyResponse> HandleDbDropAndRecreate(
    LambdaRequest req,
    Res res,
    AuthContext auth)
  {
    var deny = Auth.RequireSuperAdmin(auth, res);
    if (deny is not null) return deny;

    var required = Environment.GetEnvironmentVariable("MIGRATE_SECRET") ?? string.Empty;
    if (!string.IsNullOrEmpty(required))
    {
      var got = Validation.GetHeader(req, "x-migrate-secret") ?? string.Empty;
      if (!string.Equals(got, required, StringComparison.Ordinal)) return res.Forbidden("Bad migrate secret");
    }

    if (!string.Equals(Environment.GetEnvironmentVariable("PGDATABASE") ?? string.Empty, "postgres", StringComparison.Ordinal))
    {
      return res.BadRequest("CONFIG_ERROR", "PGDATABASE must be 'postgres' to drop/create databases");
    }

    var name = (req.Query.TryGetValue("name", out var n) ? n : string.Empty).Trim();
    if (string.IsNullOrEmpty(name)) return res.BadRequest("BAD_REQUEST", "Missing query param: ?name=");
    if (!Validation.IsValidDbName(name)) return res.BadRequest("BAD_REQUEST", "Bad database name");
    if (name is "postgres" or "rdsadmin") return res.BadRequest("BAD_REQUEST", "Refusing to drop system database");

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null) return res.BadRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

    // terminate existing connections
    const string terminateSql = """
      SELECT pg_terminate_backend(pid)
      FROM pg_stat_activity
      WHERE datname = $1
        AND pid <> pg_backend_pid();
      """;
    await ExecuteAsync(conn, null, terminateSql, [name]);

    try
    {
      await ExecuteAsync(conn, null, $"DROP DATABASE IF EXISTS {name} WITH (FORCE);", []);
      await ExecuteAsync(conn, null, $"CREATE DATABASE {name};", []);
      return res.Ok(new { ok = true, recreated = name });
    }
    catch (Exception ex)
    {
      return res.BadRequest("DB_RECREATE_FAILED", ex.Message);
    }
  }
}
